src/mlpack/methods/decision_tree/decision_tree_impl.hpp

```diff
@@ -56,6 +56,98 @@ DecisionTree<FitnessFunction,
   Train(data, labels, numClasses, minimumLeafSize);
 }
 
+//! Construct and train.
+template<typename FitnessFunction,
+         template<typename> class NumericSplitType,
+         template<typename> class CategoricalSplitType,
+         typename ElemType,
+         bool NoRecursion>
+template<typename MatType>
+DecisionTree<FitnessFunction,
+             NumericSplitType,
+             CategoricalSplitType,
+             ElemType,
+             NoRecursion>::DecisionTree(MatType&& data,
+                                        const data::DatasetInfo& datasetInfo,
+                                        arma::Row<size_t>&& labels,
+                                        const size_t numClasses,
+                                        const size_t minimumLeafSize)
+{
+  // move the data to avoid copy
+  MatType tmpData = std::move(data);
+  arma::Row<size_t> tmpLabels = std::move(labels);
```

Comment:
Yeah, I think it would be nice to adapt some of the mlpack code to use universal references like this, but there are a few problems that can make it tricky in some places:

 * universal references will match anything, so often some SFINAE may need to be done in order to restrict the matching to, e.g., only Armadillo types or something like this
 * since the universal reference requires the type to be deduced for the method itself, then you have to actually have a separate template parameter for the input matrix; i.e. the `NeighborSearch` class has a template parameter `MatType`, but you can't do this:

```cpp
template<..., typename MatType, ...>
class NeighborSearch
{
  NeighborSearch(MatType&& data, ...); // not a universal reference
};
```

but instead you have to do something like

```cpp
template<..., class MatType, ...>
class NeighborSearch
{
  template<typename MatInType>
  NeighborSearch(MatInType&& m, ...); // universal reference, but also requires SFINAE to restrict to Armadillo types
};
```

There may be some other minor concerns I haven't thought of too.  Anyway, I do agree, it would be nice to look through the mlpack codebase again and see what we can adapt to use universal references.  I think if I have time, I might start by looking at everything in `src/mlpack/core/tree/` to start with; the tree constructors could definitely be condensed using universal references.